enhancements/kubelet/master-bootstrap-credentials.md

```diff
@@ -0,0 +1,116 @@
+---
+title: master-bootstrap-credentials
+authors:
+  - "@deads2k"
+reviewers:
+  - "@rphillips"
+approvers:
+  - "@rphillips"
+creation-date: 2020-01-31
+last-updated: 2020-01-31
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Master Bootstrap Credentials
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+Update the master bootstrap credentials used to request the initial CSRs for a kubelet.
+Requirements for such a credential:
+ 1. long lived.  It should not expire on short time frames.
+ 2. revokable.  We must be able to revoke this credential if it is compromised.
+
+## Motivation
+
+In order to scale masters horizontally, replace masters, and shutdown and restore clusters,
+we need to have master bootstrap credentials that last longer.
+On workers, we use service account credentials
+
+### Goals
+
+Make it possible to
+ 1. scale masters horizontally
+ 2. replace masters
+ 3. shutdown and restore clusters
+
+### Non-Goals
+
+Actually perform the scaling.
+We need to remove this particular impediment, but the rest of the effort is out of scope.
+
+## Proposal
+
+ 1. In the installer, place the bundle from `kubelet-bootstrap-kubeconfig-ca-bundle` into openshift-config
+    called `configmap/kubelet-bootstrap-kubeconfig`.
+ 2. In the installer, sign the `kubelet-client` for ten years.  This is the actual bootstrap credential.
+ 3. In the cluster-kube-apiserver-operator, union in the content from `configmap/kubelet-bootstrap-kubeconfig`.
+ 4. In the MCO, if the `configmap/kubelet-bootstrap-kubeconfig` is deleted, use the "normal" service account token
+    logic that provides bootstrap authentication credentials for workers.
+ 5. The MCO would rollout the new machine configs and the kube-apiserver would effectively "revoke" the long lived client certificates.
+
+### Background about how kubelets start up
+
+ 1. If a current kubelet client certificate exists and is valid, use it and skip the rest.
+ 2. Use the bootstrap kubelet credentials to create a client CSR on the kube-apiserver.
+ 3. machine-approver (or cluster-admin) approvers the CSR created by the kubelet's bootstrap credentials.
+ 4. Kubelet uses bootstrap kubelet credentials to retrieve the signed certificate from the client CSR.
+ 
+If the bootstrap kubelet credentials are not valid (for instance, an expired certificate), then the kubelet is unable to create
+the CSR on the kube-apiserver and the kubelet remains unable to join the cluster.
+
+### Test Plan
+
+This is left to @rphillips who promises to write a good one.
```

Comment:
LOL.

We can replace that with:

* Unit tests within MCO validating the deletion of the configmaps and the creation of the bootstrap kubeconfig onmasters
* e2e test within MCO testing the functionality within the cluster

Reply:
I can also add more detail in the test plan